### ISSUE TITLE
Consider adding `module-info.class`

### ISSUE THREAD

**Issue by ascopes** [mockito/mockito#2604]
Would it be possible to bundle a Java-9-compiled `module-info.class` in with mockito-inline using the multi-version JAR feature (i.e. placing the class in `META-INF/versions/9/module-info.class`).

Although Mockito already distributes as an automatic module, the requirements on bytebuddy and the bytebuddy agent do not get resolved automatically, which results in a somewhat cryptic error the first time you try to use mocks.

I believe that the following would resolve this issue:

```java
module org.mockito {
  requires transitive net.bytebuddy;
  requires transitive net.bytebuddy.agent;

  exports org.mockito;
  exports org.mockito.{other packages here};
}
```

This would reduce the confusion around JPMS for new users, and remove the need to directly depend on implementation detail.

**Comment by TimvdLippe** [comment]
Yes I think that makes sense. All packages except anything in `org.mockito.internal` are implicitly exported.

Do you mind sending us a PR adding the necessary files? Thanks!

**Comment by ascopes** [comment]
@TimvdLippe at the moment I seem to be having trouble getting gradle to enable the `module-info.java` compilation path to see the other packages that are present.

This might be due to my lack of experience working with Gradle (since I tend to use Maven where this is much simpler to implement). Is there any specific examples of integrations with this, or the way that you think would be best to implement this?

**Comment by TimvdLippe** [comment]
Hm, I am not sure, as we are using typical stock Gradle. Would it maybe be better if we have hand-written `module-info.java` files that we then check during build to ensure they are correct?

**Comment by ascopes** [comment]
That could work, I think the main issue is getting it to compile under a different target to the rest of the classes though, without extracting it to a separate module.

The other alternative could be to amend the Automatic-Module-Name to something like org.mockito.internal and then have a separate JAR that imports this and any other dependencies. That may be a breaking change to anyone relying on the existing module name though.

**Comment by TimvdLippe** [comment]
I am not 100% sure what the impact of the `module-info` files are. I know JUnit5 ships them and I think for those users that want to consume Mockito with proper modules, we should ship these files.

For me, the following criteria are necessary for us to ship these files:
1. We need to have a regression test (can be a subproject in https://github.com/mockito/mockito/tree/main/subprojects) that fails if we erroneously omit a module import/export in one of our files
2. We need to make these files maintainable. Preferably we automated it, but if that's not possible, it should be straightforward to update these files
3. Existing users that use modules for other dependencies should be able to opt-in (to avoid a breaking change). Alternatively, we can include this in Mockito 5, which is scheduled for release in a couple of months
4. For those users that do not use modules, there should not be any observable runtime changes

Do you think we can make that work?

**Comment by ascopes** [comment]
The only real impact is that it adds an extra layer of encapsulation. By default it will just behave as if it depends on nothing and exports all packages publically by using Automatic-Module-Name in the manifest.

For those not using modules, they will just behave in the same way, so should see no observable differences other than packages _not_ exported by Mockito explicitly will become inaccessible. 

Existing users using modules will already have to require Mockito directly, just by including module-info rather than using Automatic Modules, it allows Mockito to explicitly state that it needs ByteBuddy.

**Comment by xenoterracide** [comment]
so what's going on with this? at minimum could we get an automatic module? I might try to do a PR, I need to look at the other issue.

**Comment by xenoterracide** [comment]
I've started work on this PR, but javadoc doesn't want to build with the existing exclude.

How would you feel about updating the java toolchain to 17 and forcing javacompile/test to use java 11 with -release? I'm thinking I can fix the remaining javadoc issues by using `@snippet` instead of html. Can then verify that the output is still correct. I don't *think* this approach causes any real world issues.

**Comment by xenoterracide** [comment]
So unfortunately I'm stumped, I left comments in the PR but here's the short of what I've figured out.

line 180 of `InlineDelegateByteBuddyMockMaker` is throwing java.lang.ClassNotFoundException: org.mockito.internal.creation.bytebuddy.inject.MockMethodDispatcher, I imagine this isn't the only problem/like this, but I don't understand this code well enough to figure out how to patch it. If someone has time to help, maybe we can achieve the goal.

Some, but probably not all internal packages will have to be exported if not refactoring for this, as jars like the jupiter extension depend on `internal` packages. Since a module is 1 to 1 with a jar, these packages might be internal to the org.mockito namespace, but they are not to jpms.

The state of the PR is I have exported/opened everything because of the `InlineDelegateByteBuddyMockMaker` issue. I'd rather play clean up later, get working first.
